acurl/src/response.pyx

```diff
@@ -0,0 +1,224 @@
+from cpython cimport array
+from json import loads
+from cpython.list cimport PyList_New
+from libc.stdio cimport printf
+
+cdef struct BufferNode:
+    size_t len
+    char *buffer
+    BufferNode *next
+
+cdef list acurl_extract_cookielist(CURL* curl):
+    cdef void* start_raw
+    cdef curl_slist *start
+    cdef curl_slist *node
+    cdef int len = 0
+    cdef int i = 0
+    acurl_easy_getinfo_voidptr(curl, CURLINFO_COOKIELIST, &start_raw)
+    start = <curl_slist*>start_raw
+    node = start
+    cdef lst = []
+    while node != NULL:
+        lst.append(node.data.decode("UTF-8"))
+        node = node.next
+    curl_slist_free_all(start)
+    return lst
+
+cdef class _Response:
+    cdef BufferNode* header_buffer
+    cdef BufferNode* header_buffer_tail
+    cdef BufferNode* body_buffer
+    cdef BufferNode* body_buffer_tail
+    cdef CURL* curl
+    cdef Session session
+    cdef object future
+    cdef readonly unsigned long start_time
+    cdef readonly Request request
+    cdef _Response _prev
+
+    def __cinit__(self):
+        # Technically it's dangerous to leave curl and session members
+        # uninitialized, but we hope no one calls our init method directly...
+        self.header_buffer = NULL
+        self.header_buffer_tail = NULL
+        self.body_buffer = NULL
+        self.body_buffer_tail = NULL
+        self._prev = None
+
+    def __dealloc__(self):
+        cdef BufferNode* ptr
+        cdef BufferNode* old_ptr
+        ptr = self.header_buffer
+        while ptr != NULL:
+            old_ptr = ptr
+            ptr = ptr.next
+            free(old_ptr)
+        ptr = self.body_buffer
+        while ptr != NULL:
+            old_ptr = ptr
+            ptr = ptr.next
+            free(old_ptr)
+        # Dealloc curl
+
+    @staticmethod
+    cdef _Response make(Session session, CURL* curl, object future, unsigned long time, Request request):
+        cdef _Response r = _Response.__new__(_Response)
+        r.session = session
+        r.curl = curl
+        r.future = future
+        r.start_time = time
+        r.request = request
+        return r
+
+    cdef long get_info_long(self, CURLINFO info):
+        cdef long value
+        acurl_easy_getinfo_long(self.curl, info, &value)
+        return value
+
+    cdef str get_info_str(self, CURLINFO info):
+        cdef char* value
+        acurl_easy_getinfo_cstr(self.curl, info, &value)
+        return value.decode('UTF-8')
+
+    cdef double get_info_double(self, CURLINFO info):
+        cdef double value
+        acurl_easy_getinfo_double(self.curl, info, &value)
+        return value
+
+    cdef list get_cookielist(self):
+        return acurl_extract_cookielist(self.curl)
+
+    def _set_prev(self, prev):
+        self._prev = prev
+
+    @property
+    def status_code(self):
+        return self.get_info_long(CURLINFO_RESPONSE_CODE)
+
+    @property
+    def response_code(self):
+        return self.get_info_long(CURLINFO_RESPONSE_CODE)
```

Comment:
Having two identical methods with different names is a leftover from the old acurl API, but ideally we'd deprecate one (i.e. make it raise a deprecation warning, and then eventually remove it).  The `requests` library that is our API inspiration only has `status_code` so that's the one we prob want to keep.

Reply:
Good point, I'm somewhat unfamiliar with the process of deprecation in regards to this sort of thing. I've implemented a warning to be raised when the method is called, but is there any specific way in which we set deadlines etc for deprecation?

Reply:
Nothing agreed wrt timescales because we haven't really done any deprecations so far.  It might be one to ask @jb098 for his input on.  I would suggest to remove it 6 months after the deprecation is merged in (since we don't really have release versions; if we did then I would say something like "2 versions after it was merged").